### ISSUE TITLE
Will this work?

### ISSUE THREAD

**Issue by WuuBoLin** [crazywolf132/supermaven.el#1]
I've noticed that the URL path for downloading the binary has changed to `/download-path-v2`, and it seems that editor should be set to `neovim` to download.

However, after successfully downloading the binary, I'm still not receiving any suggestions.

**Comment by luciusmagn** [comment]
I am also curious

**Comment by WuuBoLin** [comment]
I had forwarded a [message](https://discord.com/channels/1217199106829848656/1224604785169268867/1304381014805581874) on the support Discord, and seems like they don't have any plans to provide an Emacs plugin.

Although [they joined Cursor recently](https://supermaven.com/blog/cursor-announcement), I still think Supermaven is still the best AI tab completion model out there, and genuinely hope to use it in Emacs.

**Comment by luciusmagn** [comment]
Yeah, I am trying to get it to work. I used the -v2 path, but activating it seems to be a struggle, hahah. Will post if I figure out something

**Comment by luciusmagn** [comment]
They changed the API!

This package sends a cursor_update message, which is no longer used!

I am trying to figure out from the neovim plugin which one is the correct one now!

**Comment by WuuBoLin** [comment]
Yes, I did notice that the `cursor_update` API is no longer used. Also did the file change API.
I tried the binary downloaded from v1 API, but it seems have some text escaping issues.

**Comment by luciusmagn** [comment]
Yeah, I found that they kind of bundled that stuff up.

Apparently messages look like this roughly:

```json

{
  "kind": "state_update",
  "newId": "123",
  "updates": [
    {
      "kind": "cursor_update",
      "path": "/path/to/file",
      "offset": 724
    },
    {
      "kind": "file_update",
      "path": "/path/to/file",
      "content": "file content"
    }
  ]
}
```

**Comment by luciusmagn** [comment]
I don't know what the newId is for, though

**Comment by WuuBoLin** [comment]
Also. I recommend you to have a peak at Zed's custom implementation. It's written in Rust so easy to understand.

**Comment by luciusmagn** [comment]
Awesome idea, thanks!

**Comment by luciusmagn** [comment]
Alright, thanks to the Zed and Nvim stuff, I was able to start sending correctly formatted messages, but it complains that it has no contents for file.

I assume there is some initial message I need to use

**Comment by WuuBoLin** [comment]
Maybe the greeting? IDK

But thank you for looking into this!

**Comment by luciusmagn** [comment]
Well, I am not sure if I see it correctly, but it seems to me like Zed doesn't send a greeting message at all

**Comment by luciusmagn** [comment]
I had a fucking typo lmao, never mind. So it seems I got that to work, however, I noticed that the format the completions have has also changed, but hey, at least we are a step farther, hahah

**Comment by luciusmagn** [comment]
Okay, I did something, hahaha, it is now doing insane things to my buffers, hahah

**Comment by super-jacob** [comment]
`/download-path-v2` will fetch the new version of the binary, which uses a different protocol from the one used by the Zed and Neovim plugins. If you're aiming to use the same protocol as the Zed or Neovim plugins, you should use the old binary path from `/download-path`. It is still supported by the backend.

The benefit of using the new protocol is that it supports jump and deletion suggestions. However, there's currently no open source plugin that can be used as a reference implementation for the new protocol.

**Comment by rolandtritsch** [comment]
@super-jacob Is there documentation that you can point us to that describes the protocol for -v1 and -v2?

**Comment by rolandtritsch** [comment]
@luciusmagn I am also lookying at this. Let's start to PR the hell out of this.

**Comment by rolandtritsch** [comment]
![irony](https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExdmdmNjNhd213YTl3ZDh1dXE2cW13azFydXgwMGViZ2o5c2JxcHc1YiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/bzbbBzmcaOF7q/giphy.gif)

I am using GitHub Copilot to develop a Supermaven plugin. The irony of it ...

**Comment by WuuBoLin** [comment]
@super-jacob, thank you for the reply! I wasn't expecting that. Finally have an idea of `v1` vs `v2`.

As @rolandtritsch mentioned, having documentation that describes the protocol for both **-v1** and **-v2** would be great, even if it only includes input and output types, or else we'll have to reverse engineer it.

_A huge thanks to everyone who has taken the time to check this out. I thought no one would care about this. LOL._

**Comment by rolandtritsch** [comment]
@crazywolf132 Can you make @luciusmagn and @rolandtritsch contributors on the repo so that we can merge PRs and move fast(er)?

@luciusmagn Let's not commit/push to main directly, but always use a PR to document the change/contribution. Agreed?

**Comment by rolandtritsch** [comment]
@luciusmagn @super-jacob @crazywolf132 There is probably a couple of things that need to get fixed. Let's create seperate issues/PRs for each and every one. Let's tackle the upgrade to sm-agent version 26 here #6.

**Comment by rolandtritsch** [comment]
Was just reading the announcement. And the discussion on discord.

Comming back to the question: Will this work?

Maybe.

Seems the plan is to keep on maintaining a version of sm-agent that can be used to build and support plugins into other editors than cursor. Let's wait and see, if and how much this becomes a reality.

But in any case without documentation developing community plugins (by reverse engineering the messages) will be hard and time-consuming (time that I personally do not have).

So ... here is my plan ...

* I will try to make #4 work and will try to make this package work again against `sm-agent` version 15
* Then I will wait for the documentation on v1 and v2 of the protocol to be released
* In the meantime I will use copilot instead of supermaven (in emacs), because I do not want to switch to cursor or another editor _just_ to get access to supermaven

**Comment by luciusmagn** [comment]
Hoo, a lot of awesome developments here! I will dig right into it! (after work)

**Comment by binbandit** [comment]
@rolandtritsch, @luciusmagn now that we have merged this big fix. We just need to look at fixing the overlay code
